pkg/proxy/index/index_controller.go

```diff
@@ -0,0 +1,294 @@
+/*
+Copyright 2021 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package index
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	tenancyinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	tenancylister "github.com/kcp-dev/kcp/pkg/client/listers/tenancy/v1alpha1"
+)
+
+const (
+	controllerName = "kcp-clusterworkspace-index"
+
+	clusterWorkspaceResyncPeriod = 2 * time.Hour
+)
+
+type Index interface {
+	Lookup(logicalCluster logicalcluster.Name) (string, bool)
+}
+
+type ClusterWorkspaceClientGetter func(shard *tenancyv1alpha1.ClusterWorkspaceShard) (kcpclientset.ClusterInterface, error)
+
+func NewController(
+	rootHost string,
+	clusterWorkspaceShardInformer tenancyinformer.ClusterWorkspaceShardInformer,
+	clientGetter ClusterWorkspaceClientGetter,
+) (*Controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
+
+	c := &Controller{
+		queue: queue,
+
+		rootHost:     rootHost,
+		clientGetter: clientGetter,
+
+		clusterWorkspaceShardIndexer: clusterWorkspaceShardInformer.Informer().GetIndexer(),
+		clusterWorkspaceShardLister:  clusterWorkspaceShardInformer.Lister(),
+
+		shardClusterWorkspaceInformers: map[string]cache.SharedIndexInformer{},
+		shardClusterWorkspaceStopCh:    map[string]chan struct{}{},
+
+		workspaceShardNames: map[logicalcluster.Name]string{},
+		shardBaseURLs:       map[string]string{},
+	}
+
+	c.clusterWorkspaceHandler = cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ws := obj.(*tenancyv1alpha1.ClusterWorkspace)
+			c.lock.RLock()
+			got := c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)]
+			c.lock.RUnlock()
+
+			if expected := ws.Status.Location.Current; got != expected {
```

Comment:
what does `ws.Status.Location.Current` hold? I checked on my local instance and it seems to be always holding `root`. Should it hold a shard name?

Reply:
yeah I think it should

Reply:
it should